go/pgprotocol/client/extended.go

```diff
@@ -227,6 +233,250 @@ func (c *Conn) PrepareAndExecute(ctx context.Context, queryStr string, params []
+func (c *Conn) Execute(ctx context.Context, portalName string, maxRows int32, callback func(ctx context.Context, result *query.QueryResult) error) (completed bool, err error) {
+	c.bufmu.Lock()
+	defer c.bufmu.Unlock()
+
+	if err := c.writeExecute(portalName, maxRows); err != nil {
+		return false, fmt.Errorf("failed to write Execute: %w", err)
+	}
+
+	if err := c.writeSync(); err != nil {
+		return false, fmt.Errorf("failed to write Sync: %w", err)
+	}
+
+	if err := c.flush(); err != nil {
+		return false, fmt.Errorf("failed to flush: %w", err)
+	}
+
+	// Process execute responses.
+	return c.processExecuteResponses(ctx, callback)
+}
+
+// argsToParams converts Go values to PostgreSQL text format parameters.
+func argsToParams(args []any) ([][]byte, error) {
+	params := make([][]byte, len(args))
+	for i, arg := range args {
+		param, err := argToParam(arg)
+		if err != nil {
+			return nil, fmt.Errorf("arg %d: %w", i, err)
+		}
+		params[i] = param
+	}
+	return params, nil
+}
+
+// argToParam converts a single Go value to PostgreSQL text format.
+func argToParam(arg any) ([]byte, error) {
+	if arg == nil {
+		return nil, nil // NULL is represented as nil
+	}
+
+	switch v := arg.(type) {
+	case string:
+		return []byte(v), nil
+	case []byte:
+		return v, nil
+	case int:
+		return []byte(strconv.FormatInt(int64(v), 10)), nil
+	case int32:
+		return []byte(strconv.FormatInt(int64(v), 10)), nil
+	case int64:
+		return []byte(strconv.FormatInt(v, 10)), nil
+	case uint32:
+		return []byte(strconv.FormatUint(uint64(v), 10)), nil
+	case uint64:
+		return []byte(strconv.FormatUint(v, 10)), nil
+	case float32:
+		return []byte(strconv.FormatFloat(float64(v), 'f', -1, 32)), nil
+	case float64:
+		return []byte(strconv.FormatFloat(v, 'f', -1, 64)), nil
+	case bool:
+		if v {
+			return []byte("true"), nil
+		}
+		return []byte("false"), nil
+	case time.Time:
+		// Use RFC3339 format which PostgreSQL understands.
+		return []byte(v.Format(time.RFC3339Nano)), nil
+	default:
+		return nil, fmt.Errorf("unsupported type: %T", arg)
+	}
+}
+
+// processExecuteResponses processes responses to an Execute command.
+// Returns true if the execution completed (CommandComplete), false if suspended (PortalSuspended).
+func (c *Conn) processExecuteResponses(ctx context.Context, callback func(ctx context.Context, result *query.QueryResult) error) (completed bool, err error) {
+	var currentFields []*query.Field
+	var batchedRows []*query.Row
+	var batchedSize int
+
+	// flushBatch sends accumulated rows via callback and resets the batch.
+	flushBatch := func() error {
+		if len(batchedRows) == 0 || callback == nil {
+			return nil
+		}
+		result := &query.QueryResult{
+			Fields: currentFields,
+			Rows:   batchedRows,
+		}
+		if err := callback(ctx, result); err != nil {
+			return err
+		}
+		batchedRows = nil
+		batchedSize = 0
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
Probably worth adding a TODO() here about the known bug

Reply:
Fixed this, and handled context cancellation properly